components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts

```diff
@@ -0,0 +1,227 @@
+    return
+  }
+
+  chrome.braveRewards.updateMediaDuration(publisherKey, data.duration)
+}
+
+const handleMediaDurationHandlerRegistrationRequest = (tabId: number, mediaType: string, data: MediaDurationHandlerRegistrationRequest) => {
+  // If we already registered a handler for this tab, exit early
+  if (mediaDurationHandlerRegistrations[tabId] && mediaDurationHandlerRegistrations[tabId].registered) {
+    return
+  }
+
+  // Mark this tab as registered
+  mediaDurationHandlerRegistrations[tabId] = {
+    registered: true
+  }
+
+  chrome.webRequest.onCompleted.addListener(
+    // Listener
+    function (details) {
+      const port = greaselionPorts[tabId]
+      if (!port) {
+        return
+      }
+      port.postMessage({
+        type: 'MediaDurationMetadataRequest',
+        url: details.url
+      })
+    },
+    // Filters
+    {
+      types: [
+        'image',
+        'media',
+        'script',
+        'xmlhttprequest'
+      ],
+      urls: [
+        data.urlRegex
+      ]
+    })
+}
+
+const getPublisherPanelInfo = (tabId: number, publisherKey: string) => {
+  chrome.braveRewards.getPublisherPanelInfo(
+    publisherKey, (result: RewardsExtension.Result, info?: RewardsExtension.Publisher) => {
+      if (result === 0 && info) {
+        rewardsPanelActions.onPublisherData(tabId, info)
+      }
+      return
+    })
+}
+
+const savePublisherInfo = (tabId: number, mediaType: string, url: string, publisherKey: string, publisherName: string, favIconUrl: string) => {
+  chrome.braveRewards.savePublisherInfo(
+    tabId,
+    mediaType,
+    url,
+    publisherKey,
+    publisherName,
+    favIconUrl,
+    (result: RewardsExtension.Result) => {
+      if (result !== 0) {
+        console.error(`Failed to save publisher info for ${publisherKey}, result is ${result}`)
+        return
+      }
+    })
+}
+
+const handleSavePublisherVisitResponse = (tabId: number, mediaType: string, data: SavePublisherVisitResponse) => {
+  if (!data.publisherKey) {
+    console.error('Failed to handle publisher visit: missing publisher key')
+    return
+  }
+
+  if (data.mediaKey && !publisherKeys[data.mediaKey]) {
+    publisherKeys[data.mediaKey] = data.publisherKey
+  }
+
+  chrome.braveRewards.getPublisherInfo(
+    data.publisherKey, (result: RewardsExtension.Result, info?: RewardsExtension.Publisher) => {
+      if (result === 0 && info) {
+        getPublisherPanelInfo(tabId, data.publisherKey)
+        return
+      }
+
+      // Failed to find publisher info corresponding to this key, so save it now
+      if (result === 9) {
+        savePublisherInfo(
+          tabId,
+          mediaType,
+          data.url,
+          data.publisherKey,
+          data.publisherName,
+          data.favIconUrl || '')
+        return
+      }
+    })
+}
+
+chrome.runtime.onMessageExternal.addListener((msg, sender, sendResponse) => {
```

Comment:
is it possible to validate sender.id here (https://developer.chrome.com/extensions/runtime#type-MessageSender) to make sure these messages are only coming from our extension?

Reply:
My understanding is that Greaselion dynamically generates an extension wrapper for its content scripts, so we can't rely on the extension ID remaining the same.

Does that sound right @pilgrim-brave? If so, is there any other way for us to guard against this? It might be nice if a Greaselion script could specify a "static" extension ID to use for a particular content script (specified in the `Greaselion.json` file). Not sure if that's feasible, though.

Reply:
Correct, that is the current behavior. It was thought that the fact that all Greaselion scripts are implemented as (privileged) component extensions would be sufficient for validation. If that is not sufficient, we can discuss solutions.

Reply:
My understanding is that any extension can send a message that would be received by a `chrome.runtime.onMessageExternal` listener. We should validate that this message comes from Greaselion if that's the case.